### PR TITLE
Doc(Migration): Add Helm v2 to v3 migration doc

### LINKED ISSUE #5398: Docs: write Helm 2 -> Helm 3 migration docs
We should have documentation available that provides a smooth migration path for Helm 2 users to familiarize what's been changed in Helm 3.

CC'ing @bridgetkromhout as I know you mentioned this to me during a call about what we should do with the docs.

related: #5397

Partial-Closes: #5892

### PR DESCRIPTION
**Pull request by hickeyma**
**What this PR does / why we need it**:
Adds user documentation for migration from Helm v2 to v3.

Partial-Closes: #5892

**Special notes for your reviewer**:

**If applicable**:
- [x] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by bacongobbler** [comment]
Would you mind taking a look at the existing FAQ and see if this applies? https://github.com/helm/helm/blob/dev-v3/docs/faq.md#changes-since-helm-2

**Comment by hickeyma** [comment]
> Would you mind taking a look at the existing FAQ and see if this applies? 

@bacongobbler Sure, I will do. Thanks.

**Comment by hickeyma** [comment]
Have a few questions wrt v3:

- Will v3 have configuration and if so will it be the same directory as v2? Will still need to do a `helm init`?
- Is this correct: v2 uses "ConfigMaps" under Tiller namespace and ownership. v3 uses "Secrets" in user namespace and Helm ownership.
- Is the release object in v3:
    - Secret resource object
    - Incremental like v2
    - Release data a base-64 encoded, gzipped archive of the entire release record (like v2)

**Comment by hickeyma** [comment]
@bacongobbler FYI, there is very little in the FAQ at the moment.

**Comment by bacongobbler** [comment]
Yes. I've been working on the blog post about changes from Helm 2 to Helm 3. I was planning on porting that over to the docs. See part 1 [here](https://helm.sh/blog/helm-3-preview-pt1/)

**Comment by bacongobbler** [comment]
Hey @hickeyma, now that #5718 has been merged, is there things inside this PR that should be added to that doc or is this safe to close?

**Comment by hickeyma** [comment]
Hey @bacongobbler. I think this doc is a standalone doc that will be used by users when migrating to v3. #5718 provides a description of the changes in v3. This doc provides the steps on how to migrate to v3.

**Comment by hickeyma** [comment]
Opening this for review now. This doc will probably develop into better user documentation but lets start it from here.

File: docs/v2_v3_migration.md [hunk 1]
```diff
@@ -11,58 +11,93 @@ Migration from v2 to v3 Helm
...
 2. Deploying a new chart:
    - Use v3 to deploy
  
-3. Upgrading an existing release:
-   - Deploy as new on v3 and delete the existing release on v2
+3. Upgrading an existing release from v2 to v3:
+   - The choices are as follows:
+     - Lose the release history: Deploy as new using v3 and delete the existing release using v2
+     - Maintain the release history: TBD: Detail the steps involved in migration of v2 history to v3. This should describe the steps manually and then any tools provided for those manual steps. Are the steps involved:
+       - Retrieve ConfigMaps for Tiller owner
```
Comment (rimusz) [inline docs/v2_v3_migration.md]: v2 releases could also be stored in secrets

Comment (hickeyma) [reply docs/v2_v3_migration.md]: Good point, thanks

Comment (hickeyma) [reply docs/v2_v3_migration.md]: Sure. Updated

**Comment by hickeyma** [comment]
Updated following beta changes and community feedback. Ready for review again.

File: docs/v2_v3_migration.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+   - State now stored as in-cluster Secrets and release object metadata has changed
+   - State persisted on a namespace basis and no longer global (release names now unique only in a namespace) 
+2. Chart repository updated:
+   - Support added for upload of a chart to a repo (`helm push`)
+   - helm search now supports both local repository searches and making search queries against Helm Hub
+   - Experimental support added for OCI registries (`helm chart <sub_command>`) :
+     - Authentication supported for repo login/logout (`helm registry <sub_command>`)
+     - Authorization (OAuth2) supported which enables chart upload/install access rights
+3. Chart apiVersion bumped to "v2" for following specification changes:
+   - Dynamically linked chart deopendencies moved to `Chart.yaml` (`requirements.yaml` removed and  requirements --> dependencies)
```
Comment (php-coder) [inline docs/v2_v3_migration.md]:
```suggestion
   - Dynamically linked chart dependencies moved to `Chart.yaml` (`requirements.yaml` removed and  requirements --> dependencies)
```

File: docs/v2_v3_migration.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+   - The only issue could be if Kubernetes cluster scoped resources (e.g. `clusterroles.rbac`) are defined in a chart. The v3 deployment would then fail even if unique in the namepsace as the resources would clash.
+   - Make sure to not override the v2 client binary (`helm`). Rename or use separate directory for one of the releases.
+   - v3 configuration no longer uses `HELM_HOME` and uses XDG directory specification instead. It is also created on the fly as need be. It is therefore independent of v2 configuration.
+   - Note: This use case is only recommended if you intend to phase out Helm v2 gradually and do not require v3 to manage any releases deployed by v2. All new releases being deployed are performed by v3 and existing v2 deployed releases are updated/removed by v2 only.
+ 
+2. Migrating Helm v2 to Helm v3:
+   - Migrating Helm v2 configuration (`helm home`) to Helm v3:
+     - - Proposal is a standalone migration tool that migrates from Helm v2 to Helm v3: https://github.com/helm/helm/issues/6154
```
Comment (php-coder) [inline docs/v2_v3_migration.md]:
```suggestion
     - Proposal is a standalone migration tool that migrates from Helm v2 to Helm v3: https://github.com/helm/helm/issues/6154
```

File: docs/v2_v3_migration.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+   - `test-failure` target removed. Use `test-success` instead
+   - Commands removed/replaced/added:
+       - chart: command consists of multiple subcommands to interact with charts and registries. Subcommands as follows:
+         - export: export a chart to directory
+         - list: list all saved charts
+         - pull: pull a chart from remote
+         - push: push a chart to remote
+         - remove: remove a chart
+         - save: save a chart directory  
+       - delete --> uninstall : removes all release history by default (previous;y needed `--purge`)
```
Comment (php-coder) [inline docs/v2_v3_migration.md]:
```suggestion
       - delete --> uninstall : removes all release history by default (previously needed `--purge`)
```

File: docs/v2_v3_migration.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+   - Go import path is changed from `k8s.io/helm` to `helm.sh/helm` 
+   - Helm SDK updated. It is no longer compatible with Helm 2 library
+
+## Migration Use Cases
+
+The migration use cases are as follows:
+
+1. Running Helm v2 and v3 concurrently on the same cluster:
+   - v2 and v3 release (history) storage are independent of each other. The changes includes the Kubernetes resource for storage and the release object metadata contained in the resource. Releases will also be on a per user namespace instead of using the the Tiller namespace (for example, v2 default Tiller namespace kube-system).v2 uses "ConfigMaps" or "Secrets" under Tiller namespace and `TILLER`ownership. v3 uses "Secrets" in user namespace and `helm` ownership. There should be no conflicts.  Releases are incremental in both v2 and v3.
+   - The only issue could be if Kubernetes cluster scoped resources (e.g. `clusterroles.rbac`) are defined in a chart. The v3 deployment would then fail even if unique in the namepsace as the resources would clash.
```
Comment (php-coder) [inline docs/v2_v3_migration.md]:
```suggestion
   - The only issue could be if Kubernetes cluster scoped resources (e.g. `clusterroles.rbac`) are defined in a chart. The v3 deployment would then fail even if unique in the namespace as the resources would clash.
```

File: docs/v2_v3_migration.md (hunk 2 above)
Comment (hickeyma) [reply docs/v2_v3_migration.md]: done

File: docs/v2_v3_migration.md (hunk 4 above)
Comment (hickeyma) [reply docs/v2_v3_migration.md]: done

File: docs/v2_v3_migration.md (hunk 5 above)
Comment (hickeyma) [reply docs/v2_v3_migration.md]: done

File: docs/v2_v3_migration.md (hunk 3 above)
Comment (hickeyma) [reply docs/v2_v3_migration.md]: Sentence no longer applies.

**Comment by hickeyma** [comment]
Thanks for review comments @php-coder. Updated now.

**Comment by rimusz** [comment]
LGTM

**Comment by hickeyma** [comment]
Closing this PR as I have created PR https://github.com/helm/helm-www/pull/287 in helm-www where docs reside now.

**Pull request closed without merging**